src/quixstreams_extensions/sinks/gcp.py

```diff
@@ -0,0 +1,127 @@
+        value: Optional[Callable[[SinkItem], str]] = None,
+        value_serializer: Optional[Callable[[Any, SerializationContext], dict]] = None,
+    ):
+        super().__init__()
+        self._db = client or firestore.Client()
+        self._collection = (
+            collection if isinstance(collection, CollectionReference) else self._db.collection(collection)
+        )
+        self._key = key or attrgetter("key")
+        self._key_serializer = key_serializer
+        self._value = value or attrgetter("value")
+        self._value_serializer = value_serializer
+
+    def write(self, batch: SinkBatch):
+        db_batch = self._db.batch()
+        for item in batch:
+            ctx = SerializationContext(batch.topic, headers=item.headers)
+            key = self._key_serializer(self._key(item), ctx) if self._key_serializer else self._key(item)
+            value = self._value_serializer(self._value(item), ctx) if self._value_serializer else self._value(item)
+
+            db_batch.set(self._collection.document(key), value)
+        db_batch.commit()
+
+
+class _NestedGoogleFirestoreSink(BatchingSink):
+    """
+    Experimental. May lead to high memory usage.
+
+    A comprehensive sink that allows putting data into a nested tree-like structure.
+    May perform more Firestore writes than incoming messages due to serving nested structure creation.
+    Doesn't perform any Firestore reads.
+
+    Example:
+        Imagine you have a SinkItem like:
+
+        {
+           "key": {"user_id": 3640832, 'store_number': 123},
+           "value": {"balance": 100},
+           "timestamp": 1720828800000,
+           "headers": None
+        }
+
+        And you would like to keep a logical structure of your data, like:
+
+        stores -> "123"
+                  daily -> "2024-07-13"
+                           balance-per-user -> "3640832" = {"balance": 100}
+
+        All you need is to define **keys_structure**:
+
+        _NestedGoogleFirestoreSink(
+            [
+                ("stores", get_store_key),
+
+                ("daily", get_day_key),
+
+                ("balance-per-user", get_user_key),
+            ]
+        )
+
+        Where `get_store_key`, `get_day_key` and `get_user_key` are Callable[[SinkItem], str]
+    """
+
+    def __init__(
+        self,
+        keys_structure: List[Tuple[str, Callable[[SinkItem], str]]],
+        client: Optional[firestore.Client] = None,
+        value: Optional[Callable[[SinkItem], str]] = None,
+        value_serializer: Optional[Callable[[Any, SerializationContext], dict]] = None,
+    ):
+        super().__init__()
+        self._keys_structure = keys_structure
+        self._db = client or firestore.Client()
+        self._value = value or attrgetter("value")
+        self._value_serializer = value_serializer
+        self._existing_nodes_cache: Dict[str, DocumentReference] = {}  # TODO: memory overflow, rocksdb needs to be here
```

Comment:
That is the place where I really would like to get access to `quixstreams.state.types.State` instance, or somewhere else, e.g. during `.flush()`